### ISSUE TITLE
swarm join with Etcd fails

### ISSUE THREAD

**Issue by smothiki** [docker/swarm#721]
I have tried swarm join with the following  etcd protocol syntax only the last syntax worked. 
```
docker run --name swarm-node swarm join --addr=172.17.8.103:2375 etcd://172.17.8.100/mod/swarm
time="2015-05-05T16:29:10Z" level=fatal msg="501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]" 
```
```
$ docker run --name swarm-node swarm join --addr=172.17.8.103:2375 etcd://172.17.8.100,172.17.8.101,172.17.8.102,172.17.8.103/mod/swarm
time="2015-05-05T16:33:36Z" level=fatal msg="501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]"
```
```
$ docker run --name swarm-node swarm join --addr=172.17.8.103:2375 etcd://172.17.8.100:4001,172.17.8.101:4001,172.17.8.102:4001,172.17.8.103:4001/mod/swarm
time="2015-05-05T16:40:12Z" level=fatal msg="501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]"
```
```
$ docker run --name swarm-node swarm join --addr=172.17.8.103:2375 etcd://172.17.8.100:4001/mod/swarm
time="2015-05-05T16:32:44Z" level=info msg="Registering on the discovery service every 25 seconds..." addr="172.17.8.103:2375" discovery="etcd://172.17.8.100:4001/mod/swarm" 
```

172.17.8.100 is the leader host. Let me know if I'm missing any details while doing swarm join

**Comment by abronan** [comment]
Hi @smothiki, I encountered this same issue last week using the etcd docker image (`coreos/etcd`), with the old and new version of etcd (`0.4.6` and `2.0`). The only way for me to test the discovery locally was to run a single local instance running the `etcd` binary directly as a standalone server. (Note: even running vagrant-etcd which setups a local cluster instance was returning the error)

So I think it's related to **etcd** somehow (coreos/etcd#509, GoogleCloudPlatform/kubernetes#3853, GoogleCloudPlatform/kubernetes#3852)

How much memory do you assign to your etcd instances? It seems that other etcd instances are OOM'ing and that only the leader is up in your case. In my case the leader was probably OOM'ing while followers were up. Try to assign more memory to your containers or VM setup, that might probably help although I'm not entirely sure.

**Comment by abronan** [comment]
**Note**: I was not even running `swarm join` but just using `curl` against the etcd API.

**Comment by smothiki** [comment]
Okay I'm running swarm on Core-OS latest stable release. I'll dig deep on the error

**Comment by buth** [comment]
I'm having a similar problem in that I can only get the join to work if I specify the current etcd cluster leader. Perhaps the issue is the client never runs [SyncCluster](https://godoc.org/github.com/coreos/go-etcd/etcd#Client.SyncCluster) after it's initialized?

**Comment by aluzzardi** [comment]
Tried to find some more documentation regarding `SyncCluster` but it's a bit sketchy.

- Should we call it when we initialize the store?
- Before every operation?
- Periodically in a goroutine?

**Comment by aluzzardi** [comment]
@smothiki Can you try against master? Should have been fixed by #782

**Comment by abronan** [comment]
There are two problems here I guess:

- The first one being the `--addr` and `--bind-addr` flags being misconfigured which completely hinders the `Sync` process. (see: https://groups.google.com/forum/#!topic/confd-users/WM4a4LfPLaY)
- The second may be related to calling `SyncCluster` with the client. (hopefully fixed by #782, but the questions of whether we should call it before each operations remains open)

**Comment by kelseyhightower** [comment]
@aluzzardi @abronan I'm going to get someone from CoreOS to review this, but I think we might want to consider making a requirement that people use etcd 2.x for swarm discovery.

Recent changes in go-etcd may also help with some of the issues listed in this issue.

**Comment by xiang90** [comment]
@aluzzardi If you know your cluster configuration is changing periodically, then you need to call syncCluster periodically accordingly.

**Comment by kelseyhightower** [comment]
@aluzzardi To be clear, cluster configurations in this case would be adding and removing etcd members. Leader elections should not require a cluster sync.

**Comment by abronan** [comment]
@kelseyhightower @xiang90 Thank you for the clarification. As of now we use `v0.4.6` for the integration tests, we might switch to `v2.0.6` then. Regarding the sync it is safe then that we add the periodic `syncCluster` for those who manage their own Etcd cluster. A 10 minute refresh sounds reasonable compared to the cost of having an outdated configuration in the long run.

**Comment by aluzzardi** [comment]
@kelseyhightower thanks for chiming in, we really appreciate it.

Given the features in our roadmap, moving forward we're going to use `etcd` for a lot more than just discovery (persistent state storage and leader election are currently being worked on) so it's great if we can get feedback from your side (for instance, make etcd 2.x a requirement as you mentioned).

If you don't mind I'll ping you on relevant PRs going forward

**Comment by xiang90** [comment]
@aluzzardi I am happy to discuss more about etcd with you guys. We also plan to make etcd to store more data and be more reliable. 

We have a great roadmap to work on. It would be great to listen your use case and requirements.

**Comment by aluzzardi** [comment]
@smothiki Could you retry with `master`?

We have:
- Updated our integration tests to run against `v2.0.11` #801
- Updated our go-etcd library #801
- Called `SyncCluster` #782

**Comment by smothiki** [comment]
Sure. I'll retry with latest code and update in the comments

**Comment by abronan** [comment]
@aluzzardi @smothiki Just a reminder: I removed the call to `SyncCluster` in #801 because this was breaking integration tests after updating `etcd` to `v2.0.11`. Please tell us if something goes wrong so we can fix this ASAP.

**Comment by smothiki** [comment]
Just tested the syntax is working with just etcd IP

**Comment by abronan** [comment]
@smothiki So this still does not work with multiple IPs right?

**Comment by smothiki** [comment]
Just tried with the latest swarm code and its not working.
```
docker run --name swarm-node swarm join --addr=172.17.8.100:2375 etcd://172.17.8.100,172.17.8.101/mod/swarm
time="2015-05-27T03:32:05Z" level=info msg="Registering on the discovery service every 20s..." addr="172.17.8.100:2375" discovery="etcd://172.17.8.100,172.17.8.101/mod/swarm" 
time="2015-05-27T03:32:05Z" level=error msg="501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]" 
```
```
docker run --name manage -p 2395:2375 swarm manage  etcd://172.17.8.100,172.17.8.101/mod/swarm
time="2015-05-27T01:14:20Z" level=info msg="Listening for HTTP" addr=":2375" proto=tcp 
time="2015-05-27T01:14:20Z" level=error msg="Discovery error: 501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]" 
time="2015-05-27T01:14:20Z" level=error msg="Discovery error: Unexpected watch error" 
```
```
docker -H 172.17.8.102:2395 info
Containers: 0
Strategy: spread
Filters: affinity, health, constraint, port, dependency
Nodes: 0
```

**Comment by abronan** [comment]
Thanks @smothiki. I will investigate on this. Which version of `etcd` were you using? We recently switched to `v2.0.11` (latest) but the call to `SyncCluster` seems to panic for some reasons.

That would be nice if you could also provide us with the complete command you used to launch `etcd`. That would help to solve the issue with `SyncCluster`.

**Comment by smothiki** [comment]
I'm using etcd 0.4.9 which comes with latest core-os release 
https://coreos.com/releases/#647.0.0

**Comment by abronan** [comment]
@smothiki we pushed a fix in #867 in time for the `rc2`. Just tested with a fresh `etcd v2.0.11` cluster and it works fine. I haven't tested on older versions of etcd though.

**Comment by abronan** [comment]
I'm closing this. @smothiki Let us know if you encounter any other related issue and thanks again for reporting!
